File: web/src/lib/retag.ts
// Editing a track's tags in place: the language, and the flags the rules read.
// The service rewrites the Matroska header with mkvpropedit and judges the
// file again, so the sheet reloads to the verdict the tag change earns.

import { request } from '$lib/api';
import type { LibraryFile, Track } from '$lib/library';

// The flags on offer, per kind. The service reads each on one kind: a
// commentary 2.0 must not count as the stereo track, and forced and SDH are
// what the subtitle rules go by.
export type Flag = 'commentary' | 'forced' | 'sdh';

export type FlagOffer = { name: Flag; label: string; hint: string };

export const FLAGS: Record<string, FlagOffer[]> = {
	audio: [
		{
			name: 'commentary',
			label: 'Commentary',
			hint: 'Not the main mix. Never stands in for the stereo track, and the commentary rule may drop it.'
		}
	],
	subtitle: [
		{
			name: 'forced',
			label: 'Forced',
			hint: 'Shown even with subtitles off. The SDH rule never drops it.'
		},
		{
			name: 'sdh',
			label: 'SDH',
			hint: 'For the deaf and hard-of-hearing. The SDH rule drops it where the same language keeps a full subtitle.'
		}
	]
};

// What to set on every track named. A language left out is left alone; only
// the flags named move.
export type Edit = { lang?: string; flags?: Partial<Record<Flag, boolean>> };

// One track: the file it is in and its stream index there.
export type Target = { path: string; index: number };

// What became of one target. `verdict` is the file's status once judged again.
export type Outcome = {
	path: string;
	status: 'retagged' | 'unchanged' | 'refused' | 'failed';
	detail?: string;
	verdict?: string;
};

// The tag written for "no language", as the service spells it.
export const UNDEFINED = 'und';

/**
 * What a form set against what the track has: only what moved goes over the
 * wire, so a flag left alone stays as the file has it whatever the row's badge
 * read into it, and a flag toggled and toggled back is no edit at all.
 */
export function changes(track: Track, lang: string, flags: Partial<Record<Flag, boolean>>): Edit {
	const edit: Edit = {};
	if (lang !== (track.lang ?? UNDEFINED)) edit.lang = lang;
	const moved = (FLAGS[track.kind] ?? []).filter(
		(flag) => flags[flag.name] !== undefined && flags[flag.name] !== held(track, flag.name)
	);
	if (moved.length) {
		edit.flags = Object.fromEntries(moved.map((flag) => [flag.name, !!flags[flag.name]]));
	}
	return edit;
}

/** Whether the row reads the flag as on: the rules' reading, title included. */
export function held(track: Track, name: Flag): boolean {
	return (track.flags ?? []).includes(name);
}

export async function retagTracks(tracks: Target[], edit: Edit): Promise<Outcome[]> {
	const answer = await request<{ results?: Outcome[] }>('/api/library/retag', {
		method: 'POST',
		body: JSON.stringify({ tracks, ...edit })
	});
	return answer.results ?? [];
}

/** Whether the file can be edited in place at all: the service does Matroska
 * only, and says so itself for anything else that stops it. */
export function editable(file: { name: string }): boolean {
	return file.name.toLowerCase().endsWith('.mkv');
}

/** The tracks of one kind in a file, in order: what "the second audio track"
 * counts along. */
function ofKind(tracks: Track[], kind: string): Track[] {
	return tracks.filter((track) => track.kind === kind);
}

/**
 * Whether `other` is the same track in another file as `track` is in its own:
 * the same kind at the same place among that kind, with the same codec,
 * channels, language and title. What a season ripped alike has in every
 * episode, and what one edit is applied across.
 */
export function alike(track: Track, place: number, other: Track, otherPlace: number): boolean {
	return (
		track.kind === other.kind &&
		place === otherPlace &&
		track.codec === other.codec &&
		track.channels === other.channels &&
		track.lang === other.lang &&
		(track.title ?? '') === (other.title ?? '')
	);
}

/**
 * Every file in the title carrying a track like this one, the file it was
 * picked in first. Read off the tracks the sheet already has, which is the
 * file as last probed; the service checks each file has not changed since.
 */
export function matching(files: LibraryFile[], from: LibraryFile, track: Track): Target[] {
	const place = ofKind(from.tracks, track.kind).indexOf(track);
	if (place < 0) return [];
	const found: Target[] = [{ path: from.path, index: track.index }];
	for (const file of files) {
		if (file === from || !editable(file)) continue;
		const twin = ofKind(file.tracks, track.kind).find((other, at) =>
			alike(track, place, other, at)
		);
		if (twin) found.push({ path: file.path, index: twin.index });
	}
	return found;
}

// What a batch came to: the counts, one line saying them, and a line per file
// that could not be changed, so a hardlinked episode is named rather than
// counted.
export type Summary = { changed: number; same: number; line: string; problems: string[] };

export function summarise(outcomes: Outcome[]): Summary {
	const changed = outcomes.filter((outcome) => outcome.status === 'retagged').length;
	const same = outcomes.filter((outcome) => outcome.status === 'unchanged').length;
	const problems = outcomes
		.filter((outcome) => outcome.status === 'refused' || outcome.status === 'failed')
		.map((outcome) => `${basename(outcome.path)}: ${outcome.detail ?? outcome.status}`);
	const parts: string[] = [];
	if (changed) parts.push(`Retagged ${count(changed, 'file')}`);
	if (same) parts.push(`${count(same, 'file')} already tagged`);
	if (problems.length) parts.push(`${count(problems.length, 'file')} could not be changed`);
	return { changed, same, line: parts.join(' · ') || 'Nothing changed', problems };
}

function count(number: number, noun: string): string {
	return `${number} ${noun}${number === 1 ? '' : 's'}`;
}

function basename(path: string): string {
	return path.slice(path.lastIndexOf('/') + 1);
}

File: src/trackstarr/retag.py
"""Apply one track edit and judge the file again.

:mod:`trackstarr.mkvtag` writes the tag; this decides whether an edit aimed
from a page may be made at all, then books what it left: the file is judged
again at once so the page reloads to the new verdict, and the *arrs and the
media servers are told as a rewrite tells them. Nothing keeps a record of the
tags themselves beyond a history line saying who changed what.
"""

import logging
import re

from . import events, library, rewrites, sweep, sweep_cache
from .arr import Arr, innermost
from .langs import norm_lang
from .media_server import refresh_servers
from .mkvtag import FLAGS, UNDEFINED, Edit, Outcome, Result, edit_track, mistook, unwritable
from .policy import Policy
from .processing import Job, process
from .sweep_cache import cache_key, cache_path

log = logging.getLogger(__name__)

#: Most tracks one request may edit: a season's worth, with room. Matches the
#: files a title's sheet lists (library.MAX_FILES), which is where a batch is
#: picked from.
MAX_TARGETS = 200

#: What a language must look like once normalised, before mkvpropedit sees
#: it. mkvpropedit refuses a code that is not ISO 639, with a message worth
#: passing on, so this only turns away what could never be one.
_CODE = re.compile(r"[a-z]{2,3}")


def parse_edit(body: dict) -> Edit:
    """The edit a request body asks for. Raises ValueError with the refusal."""
    lang = body.get("lang")
    if lang is not None:
        if not isinstance(lang, str):
            raise ValueError("lang is a language code")
        # 639-2/B, as every track tag is compared: "ja" and "Japanese" land
        # as jpn, and nothing at all as und.
        code = norm_lang(lang) or UNDEFINED
        if not _CODE.fullmatch(code):
            raise ValueError(f"{lang!r} is not a language code")
        lang = code
    flags = body.get("flags") or {}
    if not isinstance(flags, dict):
        raise ValueError("flags is a map of flag name to on or off")
    known = {flag.name for flag in FLAGS}
    for name, value in flags.items():
        if name not in known:
            raise ValueError(
                f"{name!r} is not a flag; the flags are {', '.join(sorted(known))}"
            )
        if not isinstance(value, bool):
            raise ValueError(f"{name} is on or off")
    if lang is None and not flags:
        raise ValueError("nothing to change")
    return Edit(lang, dict(flags))


def refusal(path: str, stored: dict | None) -> str | None:
    """Why an edit aimed from a page cannot be made, or None.

    ``stored`` is the file's sweep cache entry, whose size and mtime say
    whether the tracks the page showed are still the file's; a file with no
    entry has shown nothing to aim an edit at. The rules need no such check,
    since they aim at a probe of their own.
    """
    if why := unwritable(path, stored):
        return why
    if stored is None:
        return "the file has no verdict yet; sweep or re-check the title first"
    return None


def apply(
    path: str,
    index: int,
    edit: Edit,
    by: str,
    stored: dict | None,
    title: library.Title | None = None,
) -> Result:
    """Make one edit to one track and judge the file again.

    ``stored`` is the file's sweep cache entry and ``title`` the library title
    it belongs to, whose original language the verdict turns on. The file is
    read back after the edit rather than trusted: the history records what
    changed, and the new verdict is booked so the sheet reloads to it rather
    than waiting on the next sweep.
    """
    if why := refusal(path, stored):
        return Result(path, Outcome.REFUSED, why)
    edited = edit_track(path, index, edit)
    if isinstance(edited, Result):
        return edited
    before, wanted, kind, after = edited
    changed = {
        name: {"from": before[name], "to": after.get(name)}
        for name in before
        if after.get(name) != before[name]
    }
    if changed:
        events.record("retagged", path=path, index=index, kind=kind, changed=changed, by=by)
        log.info("%s stream %d retagged by %s: %s", path, index, by, changed)
    # The verdict stands either way, so it is reached before the edit is held
    # against what was asked for.
    verdict = _rejudge(path, title)
    return mistook(path, wanted, after, verdict) or Result(path, Outcome.RETAGGED, "", verdict)


def _rejudge(path: str, title: library.Title | None) -> str:
    """Book the file's verdict as it now stands, and say what it is.

    Report only, through :func:`trackstarr.processing.process` so verdicts are
    reached in one place. The media servers are told too, as a rewrite tells
    them, so their own track lists follow the file's. A walk starting after
    the request's check can book over this entry at its next checkpoint; the
    edit moved the file's size and mtime, so that entry fails its own key and
    the file is probed again.
    """
    lang = title.lang if title else None
    # Keyed before the probe, as a sweep keys its verdicts, so a change landing
    # under the probe leaves a stale entry rather than a wrong one.
    key = cache_key(path, lang)
    result = process(Job(path, lang), dry_run=True)
    sweep.remember(path, key, result)
    # Our own edit moved the file, so the record follows it rather than losing
    # its claim to a change we made.
    rewrites.rekey(path, key)
    refresh_servers(path)
    return str(result.status)


def apply_all(targets: list[tuple[str, int]], edit: Edit, by: str) -> list[Result]:
    """One edit across every ``(path, stream index)`` named, in order: a
    season's untagged original track in one press. Each file is its own
    answer; one refusal stops nothing else. The *arr is asked to rescan once
    per title touched, since its media info is per title."""
    stored = sweep_cache.read(cache_path(), Policy.from_config().fingerprint())
    folders = {title.folder: title for title in library.known().titles}
    results: list[Result] = []
    # The *arr and item behind each title a file of which changed.
    rescans: dict[str, tuple[Arr, int]] = {}
    for path, index in targets:
        title = innermost(folders, path)
        result = apply(path, index, edit, by, stored.files.get(path), title)
        results.append(result)
        if result.verdict and title is not None and title.arr is not None and title.item_id:
            rescans.setdefault(title.id, (title.arr, title.item_id))
    for arr, item_id in rescans.values():
        arr.rescan(item_id)
    return results
